pkg/promtail/targets/object_reader.go

```diff
@@ -0,0 +1,236 @@
+			s3ReadLines.WithLabelValues(r.object.Key).Inc()
+			s3LogLengthHistogram.WithLabelValues(r.object.Key).Observe(float64(len(bytes)))
+
+			// send line, timestamp and labels
+			if err := r.handler.Handle(labels, time, line); err != nil {
+				level.Error(r.logger).Log("msg", "error handling line", "line", string(bytes), "error", err)
+			}
+
+			// update bytes read
+			r.object.BytesRead += int64(len(bytes))
+		}
+	}
+
+}
+
+func (r *objectReader) markPositionAndSize() {
+	r.mux.Lock()
+	defer r.mux.Unlock()
+
+	// update positions
+	positionKey := fmt.Sprintf("s3object-%s", r.object.Key)
+	positionValue := fmt.Sprintf("%s:%s:%s", strconv.FormatInt(r.object.ModifiedAt.UnixNano(), 10), strconv.FormatInt(r.object.BytesRead, 10), strconv.FormatInt(r.object.Size, 10))
+	r.positions.PutString(positionKey, positionValue)
+
+	s3ReadBytes.WithLabelValues(r.object.Key).Set(float64(r.object.BytesRead))
+	s3TotalBytes.WithLabelValues(r.object.Key).Set(float64(r.object.Size))
+}
+
+func (r *objectReader) GetObject() (*bufio.Reader, error) {
+	reader, err := r.objectClient.GetObject(context.Background(), r.object.Key)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to fetch object")
+	}
+
+	bufReader, err := r.getReader(reader)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get reader")
+	}
+
+	if err = r.setObjectSize(bufReader); err != nil {
+		return nil, errors.Wrap(err, "failed to set object size")
+	}
+
+	availableBytes, err := bufReader.Peek(int(r.object.BytesRead))
+	if err != nil {
+		level.Warn(r.logger).Log("msg", "object size is less than previous known size. object will be tailed from begining", "object", r.object.Key)
+		r.positions.Remove(fmt.Sprintf("s3object-%s", r.object.Key))
+		r.object.BytesRead = 0
+		return bufReader, nil
+	}
+
+	// skip the bytes which we have already read
+	_, err = bufReader.Discard(len(availableBytes))
+	if err != nil {
+		return nil, err
+	}
+
+	level.Info(r.logger).Log("msg", "previously known object. object will be tailed from the previous known position", "object", r.object.Key, "position", r.object.BytesRead)
+	return bufReader, nil
+}
+
+func (r *objectReader) setObjectSize(bufReader *bufio.Reader) error {
+	// read the byte so that we can get the size of the underlying buffer
+	_, err := bufReader.ReadByte()
+	if err != nil {
+		return err
+	}
+
+	// set object size
+	r.object.Size = 1 + int64(bufReader.Buffered())
+
+	// unread the byte
+	if err = bufReader.UnreadByte(); err != nil {
+		return err
+	}
+
+	return nil
+}
+func (r *objectReader) getReader(reader io.ReadCloser) (*bufio.Reader, error) {
+	// identify the file type
+	buf := [3]byte{}
+	n, err := io.ReadAtLeast(reader, buf[:], len(buf))
+	if err != nil {
+		return nil, err
+	}
+
+	rd := io.MultiReader(bytes.NewReader(buf[:n]), reader)
+	if isGzip(buf) {
+		r, err := gzip.NewReader(rd)
+		if err != nil {
+			return nil, err
+		}
+		rd = r
+	}
+
+	bufReader := bufio.NewReader(rd)
+	return bufReader, nil
+}
```

Comment:
@cyriltovena I need some help here. I am not sure If I am doing this in a proper way.